post-processing/post_processing.py

```diff
@@ -407,27 +379,45 @@ def get_display_name_info(display_name):
 
     return test_name, dict(params)
 
-def prepare_columns(columns, dni):
+def insert_key_cols(df: pd.DataFrame, index, results):
     """
-        Return a list of modified column values for a single perflog entry, after breaking up the display name column into test name and parameters. A display name index is used to determine which column to parse as the display name.
+        Modify a dataframe to include new columns (extracted from results) inserted at a given index.
 
         Args:
-            columns: str list, containing the column values for the whole perflog line, expecting the display name column format of <test_name> followed by zero or more %<param>=<value> pairs.
-            dni: int, a display name index that identifies the display name column.
+            df: dataframe, to be modified by this function.
+            index: int, index as which to insert new columns into the dataframe.
+            results: dict list, contains key-value mapping information for all rows.
     """
+    # get set of keys from all rows
+    keys = set(chain.from_iterable([r.keys() for r in results]))
+    for k in keys:
+        # insert keys as new columns
+        df.insert(index, k, [r[k] if k in r.keys() else "" for r in results])
 
-    # get display name
-    display_name = columns[dni]
-    # get test name and parameters
-    test_name, params = get_display_name_info(display_name)
+def get_axis_info(df: pd.DataFrame, axis):
+    """
+        Return the column name and label for a given axis. If a column name is supplied as units information, the actual units will be extracted from a dataframe.
 
-    # remove display name from columns
-    columns.pop(dni)
-    # replace with test name and parameter values
-    columns[dni:dni] = [params[name] for name in params]
-    columns.insert(dni, test_name)
+        Args:
+            df: dataframe, data to plot.
+            axis: dict, axis column and units.
+    """
 
-    return columns
+    # get column name of axis
+    col_name = axis.get("value")
+    # get units
+    units = axis.get("units").get("custom")
+    if axis.get("units").get("column"):
+        unit_set = set(df[axis["units"]["column"]])
+        # check all rows have the same units
+        if len(unit_set) > 1:
+            raise RuntimeError("Too many axis unit entries {0}".format(unit_set))

```

Comment:
When gathering the `unit_set`, the dataframe could likely use a `dropna()`; I noticed some logs that had rows with missing units. This won't work on anything pandas doesn't deem to be a missing value of course, and we'd have to change the `RuntimeError` check from `len(unit_set) > 1` to `len(unit_set) != 1`, but it seems reasonable to ignore missing units and type everything according to the rest of the rows.

Reply:
Yes!